vms/platformvm/validators/manager.go

```diff
@@ -164,100 +164,138 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
+		err := m.applyValidatorDiffs(subnetSet, primarySet, subnetID, diffHeight)
 		if err != nil {
 			return nil, err
 		}
 	}
 
 	// cache the validator set
-	validatorSetsCache.Put(height, vdrSet)
+	validatorSetsCache.Put(height, subnetSet)
 
 	endTime := m.clk.Time()
 	m.metrics.IncValidatorSetsCreated()
 	m.metrics.AddValidatorSetsDuration(endTime.Sub(startTime))
 	m.metrics.AddValidatorSetsHeightDiff(lastAcceptedHeight - height)
-	return vdrSet, nil
+	return subnetSet, nil
 }
 
 func (m *manager) applyValidatorDiffs(
-	vdrSet map[ids.NodeID]*validators.GetValidatorOutput,
+	targetSet, primarySet map[ids.NodeID]*validators.GetValidatorOutput,
 	subnetID ids.ID,
 	height uint64,
 ) error {
-	weightDiffs, err := m.state.GetValidatorWeightDiffs(height, subnetID)
+	// fully rebuild primary network validators at [height]
+	primaryWeightDiffs, err := m.state.GetValidatorWeightDiffs(height, constants.PlatformChainID)
 	if err != nil {
 		return err
 	}
-
-	for nodeID, weightDiff := range weightDiffs {
-		vdr, ok := vdrSet[nodeID]
-		if !ok {
-			// This node isn't in the current validator set.
-			vdr = &validators.GetValidatorOutput{
-				NodeID: nodeID,
-			}
-			vdrSet[nodeID] = vdr
-		}
-
-		// The weight of this node changed at this block.
-		if weightDiff.Decrease {
-			// The validator's weight was decreased at this block, so in the
-			// prior block it was higher.
-			vdr.Weight, err = math.Add64(vdr.Weight, weightDiff.Amount)
-		} else {
-			// The validator's weight was increased at this block, so in the
-			// prior block it was lower.
-			vdr.Weight, err = math.Sub(vdr.Weight, weightDiff.Amount)
-		}
+	for nodeID, weightDiff := range primaryWeightDiffs {
+		err := rebuildWeight(primarySet, nodeID, weightDiff)
 		if err != nil {
 			return err
 		}
-
-		if vdr.Weight == 0 {
-			// The validator's weight was 0 before this block so
-			// they weren't in the validator set.
-			delete(vdrSet, nodeID)
-		}
 	}
 
 	pkDiffs, err := m.state.GetValidatorPublicKeyDiffs(height)
 	if err != nil {
 		return err
 	}
-
 	for nodeID, pk := range pkDiffs {
-		// pkDiffs includes all primary network key diffs, if we are
-		// fetching a subnet's validator set, we should ignore non-subnet
-		// validators.
-		if vdr, ok := vdrSet[nodeID]; ok {
+		if vdr, ok := primarySet[nodeID]; ok {
 			// The validator's public key was removed at this block, so it
 			// was in the validator set before.
 			vdr.PublicKey = pk
 		}
 	}
+
+	// rebuild weights of target validators
+	targetWeightDiffs, err := m.state.GetValidatorWeightDiffs(height, subnetID)
+	if err != nil {
+		return err
+	}
+	for nodeID, weightDiff := range targetWeightDiffs {
+		err := rebuildWeight(targetSet, nodeID, weightDiff)
+		if err != nil {
+			return err
+		}
+	}
+
+	// rebuild public key of target validators, just peeking in primary validators set
+	for nodeID, vdr := range targetSet {
+		if primary, found := primarySet[nodeID]; found {
```

Comment:
Do we need this check since all subnet validators are also primary network validators

Reply:
we don't, you're right. But a bug breaking this invariant would cause a panic if the found is dropped right? I guess I am just trying to be defensive. I feel we usually do this, but happy to switch if I am wrong